File: ts/35.search-insert-position.ts
/*
 * @lc app=leetcode id=35 lang=typescript
 *
 * [35] Search Insert Position
 */

// @lc code=start
function searchInsert(nums: number[], target: number): number {
  let pivot,
    left = 0;
  let right = nums.length - 1;
  while (left <= right) {
    pivot = Math.floor(left + (right - left) / 2);
    if (nums[pivot] === target) return pivot;
    if (nums[pivot] > target) {
      right = pivot - 1;
    } else {
      left = pivot + 1;
    }
  }
  return left;
}
// @lc code=end

File: go/35.search-insert-position/35.search-insert-position.go
/*
 * @lc app=leetcode id=35 lang=golang
 *
 * [35] Search Insert Position
 */

package leetcode

// @lc code=start
func searchInsert(nums []int, target int) int {
	// do binary-search
	left, right := 0, len(nums)-1

	for left <= right {
		mid := (left + right) / 2

		if target == nums[mid] {
			return mid
		} else if target > nums[mid] {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return left
}

// @lc code=end
